x-pack/plugins/task_manager/server/plugin.ts

```diff
@@ -3,92 +3,140 @@
-        logger,
-      })
-    );
-    this.legacyTaskManager$.complete();
-
-    // we need to "drain" any calls made to the seup API
-    // before `starting` TaskManager. This is a legacy relic
-    // of the old API that should be resolved once we split
-    // Task manager into two services, setup and start, instead
-    // of the single instance of TaskManager
-    this.taskManager.then((tm) => tm.start());
+    const taskStore = new TaskStore({
+      serializer: savedObjects.createSerializer(),
+      savedObjectsRepository,
+      esClient: elasticsearch.createClient('taskManager').asInternalUser,
+      index: this.config!.index,
+      maxAttempts: this.config!.max_attempts,
+      definitions: this.definitions,
+      taskManagerId: `kibana:${this.taskManagerId!}`,
+    });
+
+    const { maxWorkersConfiguration$, pollIntervalConfiguration$ } = createManagedConfiguration({
+      logger: this.logger,
+      errors$: taskStore.errors$,
+      startingMaxWorkers: this.config!.max_workers,
+      startingPollInterval: this.config!.poll_interval,
+    });
+
+    const taskPollingLifecycle = new TaskPollingLifecycle({
+      config: this.config!,
+      definitions: this.definitions,
+      logger: this.logger,
+      taskStore,
+      middleware: this.middleware,
+      maxWorkersConfiguration$,
+      pollIntervalConfiguration$,
+    });
+    this.taskPollingLifecycle = taskPollingLifecycle;
+
+    const taskScheduling = new TaskScheduling({
+      logger: this.logger,
+      taskStore,
+      middleware: this.middleware,
+      taskPollingLifecycle,
+    });
+
+    // start polling for work
+    taskPollingLifecycle.start();
 
     return {
-      fetch: (...args) => this.taskManager.then((tm) => tm.fetch(...args)),
-      get: (...args) => this.taskManager.then((tm) => tm.get(...args)),
-      remove: (...args) => this.taskManager.then((tm) => tm.remove(...args)),
-      schedule: (...args) => this.taskManager.then((tm) => tm.schedule(...args)),
-      runNow: (...args) => this.taskManager.then((tm) => tm.runNow(...args)),
-      ensureScheduled: (...args) => this.taskManager.then((tm) => tm.ensureScheduled(...args)),
+      fetch: (opts: SearchOpts): Promise<FetchResult> => taskStore.fetch(opts),
+      get: (id: string) => taskStore.get(id),
+      remove: (id: string) => taskStore.remove(id),
+      schedule: (...args) => taskScheduling.schedule(...args),
+      ensureScheduled: (...args) => taskScheduling.ensureScheduled(...args),
+      runNow: (...args) => taskScheduling.runNow(...args),
     };
   }
+
   public stop() {
-    this.taskManager.then((tm) => {
-      tm.stop();
-    });
+    if (this.taskPollingLifecycle) {
+      this.taskPollingLifecycle.stop();
+    }
+  }
+
+  /**
+   * Ensures task manager hasn't started
+   *
+   * @param {string} the name of the operation being executed
+   * @returns void
+   */
+  private assertStillInSetup(operation: string) {
+    if (this.taskPollingLifecycle?.isStarted) {
```

Comment:
Core doesn't provide anything sadly, I discussed it with Platform last week.
The reason I went with this instead of adding a variable (I did actually have one at one point) was because they could go out of sync and it actually made sense for me for the lifecycle to be the source of truth.

It's actually a pain because the main goal here to prevent types/middle being added after the poller started... it's not **really* about the plugin lifecycle.